iroh-net/src/dns/node_info.rs

```diff
@@ -0,0 +1,267 @@
+///
+/// The domain name must either contain an _iroh_node TXT record or be a CNAME record that leads to
+/// an _iroh_node TXT record.
+pub async fn lookup_by_domain(resolver: &TokioAsyncResolver, domain: &str) -> Result<NodeAddr> {
+    let name = Name::from_str(domain)?;
+    let info = lookup_node_info(resolver, name).await?;
+    Ok(info.into())
+}
+
+/// Lookup node info by node id and origin domain name.
+pub async fn lookup_by_id(
+    resolver: &TokioAsyncResolver,
+    node_id: &NodeId,
+    origin: &str,
+) -> Result<NodeAddr> {
+    let domain = format!("{}.{}", to_z32(node_id), origin);
+    lookup_by_domain(resolver, &domain).await
+}
+
+pub(crate) async fn lookup_node_info(
+    resolver: &TokioAsyncResolver,
+    name: Name,
+) -> Result<NodeInfo> {
+    let name = ensure_iroh_node_txt_label(name)?;
+    let lookup = resolver.txt_lookup(name).await?;
+    NodeInfo::from_hickory_lookup(lookup.as_lookup())
+}
+
+fn ensure_iroh_node_txt_label(name: Name) -> Result<Name, ProtoError> {
+    if name.iter().next() == Some(IROH_NODE_TXT_LABEL.as_bytes()) {
+        Ok(name)
+    } else {
+        Name::parse(IROH_NODE_TXT_LABEL, Some(&name))
+    }
+}
+
+/// Encode a [`NodeId`] in [`z-base-32`] encoding.
+///
+/// [z-base-32]: https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt
+pub fn to_z32(node_id: &NodeId) -> String {
+    z32::encode(node_id.as_bytes())
+}
+
+/// Parse a [`NodeId`] from [`z-base-32`] encoding.
+///
+/// [z-base-32]: https://philzimmermann.com/docs/human-oriented-base-32-encoding.txt
+pub fn from_z32(s: &str) -> Result<NodeId> {
+    let bytes = z32::decode(s.as_bytes()).map_err(|_| anyhow!("invalid z32"))?;
+    let bytes: &[u8; 32] = &bytes.try_into().map_err(|_| anyhow!("not 32 bytes long"))?;
+    let node_id = NodeId::from_bytes(bytes)?;
+    Ok(node_id)
+}
+
+/// Node info contained in a DNS _iroh_node TXT record.
+#[derive(derive_more::Debug, Clone, Eq, PartialEq)]
+pub struct NodeInfo {
+    /// The node id
+    pub node_id: NodeId,
+    /// Home relay server for this node
+    #[debug("{:?}", self.relay_url.as_ref().map(|s| s.to_string()))]
+    pub relay_url: Option<Url>,
+}
+
+impl From<NodeInfo> for NodeAddr {
+    fn from(value: NodeInfo) -> Self {
+        NodeAddr {
+            node_id: value.node_id,
+            info: value.into(),
+        }
+    }
+}
+
+impl From<NodeInfo> for AddrInfo {
+    fn from(value: NodeInfo) -> Self {
+        AddrInfo {
+            relay_url: value.relay_url.map(|u| u.into()),
+            direct_addresses: Default::default(),
+        }
+    }
+}
+
+impl NodeInfo {
+    /// Create a new [`NodeInfo`] from its parts.
+    pub fn new(node_id: NodeId, relay_url: Option<Url>) -> Self {
+        Self { node_id, relay_url }
+    }
+
+    /// Convert this node info into a DNS attribute string.
+    ///
+    /// It will look like this:
+    /// `node=b32encodednodeid relay=https://myrelay.example`
+    pub fn to_attribute_string(&self) -> String {
```

Comment:
should be `as_attribute_string` according to the convention I believe

Reply:
Rereading the docs at https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv - I think `to_` is correct because this create a new owned `String`?

Reply:
you are right